lib/streamlit/runtime/partials.py

```diff
@@ -0,0 +1,181 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import hashlib
+import sys
+from functools import wraps
+from typing import (
+    Any,
+    Callable,
+    Iterator,
+    MutableMapping,
+    Optional,
+    TypeVar,
+    Union,
+    overload,
+)
+
+import cloudpickle
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+from streamlit.runtime.scriptrunner import get_script_run_ctx
+from streamlit.runtime.state.session_state_proxy import get_session_state
+
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+@overload
+def partial(
+    func: F,
+    *,
+    run_every: Optional[float] = None,
+) -> F:
+    ...
+
+
+@overload
+def partial(
+    func: None = None,
+    *,
+    run_every: Optional[float] = None,
+) -> Callable[[F], F]:
+    ...
+
+
+def partial(
+    func: Optional[F] = None,
+    *,
+    run_every: Optional[float] = None,
+) -> Union[Callable[[F], F], F]:
+    if func is None:
+        # Support passing the params via function decorator
+        def wrapper(f: F) -> F:
+            return partial(
+                func=f,
+                run_every=run_every,
+            )
+
+        return wrapper
+    else:
+        # To make mypy type narrow Optional[F] -> F
+        non_optional_func = func
+
+    print(non_optional_func.__qualname__)
+
+    @wraps(non_optional_func)
+    def wrap(*args, **kwargs):
+        ctx = get_script_run_ctx()
+        if ctx is None or len(ctx.dg_stack) == 0:
+            return
+        dg_stack = ctx.dg_stack
+        active_dg = ctx.dg_stack[-1]
+
+        # TODO(lukasmasuch): Research more on what to include in the hash:
+        h = hashlib.new("md5")
+        h.update(
```

Comment:
This is a unique ID referring to the partial based on the function name, module & delta path